api/src/opentrons/protocol_api/instrument_context.py

```diff
@@ -277,34 +287,38 @@ def dispense(
                 volume, location if location else "current position", rate
             )
         )
-        well: Optional[labware.Well]
-        last_location = self._protocol_core.get_last_location()
-
-        if isinstance(location, labware.Well):
-            well = location
-            if well.parent._core.is_fixed_trash():
-                move_to_location = location.top()
-            else:
-                move_to_location = location.bottom(
-                    z=self._well_bottom_clearances.dispense
-                )
-        elif isinstance(location, types.Location):
-            move_to_location = location
-            _, well = move_to_location.labware.get_parent_labware_and_well()
-        elif location is not None:
-            raise TypeError(
-                f"location should be a Well or Location, but it is {location}"
+        well: Optional[labware.Well] = None
+        last_location = self._get_last_location_by_api_version()
+        dispense_in_place: bool = False
+        try:
+            target = validation.validate_location(
+                location=location, last_location=last_location
             )
-        elif last_location:
-            move_to_location = last_location
-            _, well = move_to_location.labware.get_parent_labware_and_well()
-        else:
+        except validation.NoLocationError:
             raise RuntimeError(
                 "If dispense is called without an explicit location, another"
                 " method that moves to a location (such as move_to or "
                 "aspirate) must previously have been called so the robot "
                 "knows where it is."
             )
+        except validation.LocationTypeError:
+            raise TypeError(
+                f"location should be a Well or Location, but it is {location}"
+            )
+
+        if isinstance(target, validation.WellTarget):
+            well = target.well
+            if well.parent._core.is_fixed_trash():
+                move_to_location = target.well.top()
+            else:
+                move_to_location = target.location or target.well.bottom(
+                    z=self._well_bottom_clearances.dispense
+                )
```

Comment:
Slight regression in behavior: this "is trash? then use top" should only be applicable if `target.location is None`

```suggestion
            if target.location:
                move_to_location = target.location
            elif well.parent._core.is_fixed_trash():
                move_to_location = target.well.top()
            else:
                move_to_location = target.well.bottom(
                    z=self._well_bottom_clearances.dispense
                )
```